src-self-hosted/link/Wasm.zig

```diff
@@ -0,0 +1,445 @@
+    /// This section needs special handling to keep the indexes matching with
+    /// the codesec, so we cant just use a VecSection.
+    funcsec: Section,
+    /// Number of functions listed in the funcsec. Must be kept in sync with
+    /// codesec.entries.items.len.
+    funcs_count: u32,
+    codesec: VecSection,
+
+    fn init(file: fs.File, funcs_offset: u64, funcs_size: u64, code_offset: u64, code_size: u64) !Funcs {
+        return Funcs{
+            .funcsec = (try VecSection.init(spec.funcs_id, file, funcs_offset, funcs_size)).section,
+            .funcs_count = 0,
+            .codesec = try VecSection.init(spec.code_id, file, code_offset, code_size),
+        };
+    }
+
+    fn deinit(self: *Funcs) void {
+        const wasm = @fieldParentPtr(Wasm, "funcs", self);
+        self.codesec.deinit(wasm.base.allocator);
+    }
+
+    /// Add a new function to the binary, first finding space for and writing
+    /// the code then writing the typeidx to the corresponding index in the
+    /// funcsec. Returns the function index used.
+    fn new(self: *Funcs, typeidx: u32, code: []const u8) !u32 {
+        const wasm = @fieldParentPtr(Wasm, "funcs", self);
+        const file = wasm.base.file.?;
+        const allocator = wasm.base.allocator;
+
+        assert(self.funcs_count == self.codesec.entries.items.len);
+
+        // TODO: consider nop-padding the code if there is a close but not perfect fit
+        const funcidx = try self.codesec.addEntry(file, allocator, code);
+
+        if (self.funcs_count < self.codesec.entries.items.len) {
+            // u32 vector length + funcs_count u32s in the vector
+            const current = 5 + self.funcs_count * 5;
+            try self.funcsec.resize(file, current, current + 5);
+            self.funcs_count += 1;
+
+            // Update the size in the section header and the item count of
+            // the contents vector.
+            var size_and_count: [10]u8 = undefined;
+            leb.writeUnsignedFixed(5, size_and_count[0..5], 5 + self.funcs_count * 5);
+            leb.writeUnsignedFixed(5, size_and_count[5..], self.funcs_count);
+            try file.pwriteAll(&size_and_count, self.funcsec.offset + 1);
+        }
+        assert(self.funcs_count == self.codesec.entries.items.len);
+
+        var typeidx_leb: [5]u8 = undefined;
+        leb.writeUnsignedFixed(5, &typeidx_leb, typeidx);
+        try file.pwriteAll(&typeidx_leb, self.funcsec.offset + Section.header_size + 5 + funcidx * 5);
+
+        return funcidx;
+    }
+
+    fn free(self: *Funcs, funcidx: u32) void {
+        self.codesec.delEntry(funcidx);
+    }
+};
+
+/// Exports are tricky. We can't leave dead entries in the binary as they
+/// would obviously be visible from the execution environment. The simplest
+/// way to work around this is to re-emit the export section whenever
+/// something changes. This also makes it easier to ensure exported function
+/// and global indexes are updated as they change.
```

Comment:
This might be something where you would have a "exports are dirty" flag, and then write all the exports at once in the `flush` function

Reply:
That sounds like a much better way to do this indeed.

Reply:
Also consider that any given compilation, even if it is many lines of source code, will typically have a small, fixed number of exports. For example ffmpeg has hundreds of thousands of lines of source code, but the API surface is only a few dozen functions.